numba/np/arraymath.py

```diff
@@ -351,263 +351,225 @@ def array_sum_impl_axis(arr, axis):
     return impl_ret_new_ref(context, builder, sig.return_type, res)
 
 
-@lower_builtin(np.prod, types.Array)
-@lower_builtin("array.prod", types.Array)
-def array_prod(context, builder, sig, args):
-
-    def array_prod_impl(arr):
-        c = 1
-        for v in np.nditer(arr):
-            c *= v.item()
-        return c
-
-    res = context.compile_internal(builder, array_prod_impl, sig, args,
-                                   locals=dict(c=sig.return_type))
-    return impl_ret_borrowed(context, builder, sig.return_type, res)
-
-
-@lower_builtin(np.cumsum, types.Array)
-@lower_builtin("array.cumsum", types.Array)
-def array_cumsum(context, builder, sig, args):
-    scalar_dtype = sig.return_type.dtype
-    dtype = as_dtype(scalar_dtype)
-    zero = scalar_dtype(0)
-
-    def array_cumsum_impl(arr):
-        out = np.empty(arr.size, dtype)
-        c = zero
-        for idx, v in enumerate(arr.flat):
-            c += v
-            out[idx] = c
-        return out
-
-    res = context.compile_internal(builder, array_cumsum_impl, sig, args,
-                                   locals=dict(c=scalar_dtype))
-    return impl_ret_new_ref(context, builder, sig.return_type, res)
+@overload(np.prod)
+@overload_method(types.Array, "prod")
+def array_prod(arr):
+    if isinstance(arr, types.Array):
+        def array_prod_impl(arr):
+            c = 1
+            for v in np.nditer(arr):
+                c *= v.item()
+            return c
+
+        return array_prod_impl
+
+
+@overload(np.cumsum)
+@overload_method(types.Array, "cumsum")
+def array_cumsum(arr):
+    if isinstance(arr, types.Array):
+        is_integer = arr.dtype in types.signed_domain
+        is_bool = arr.dtype == types.bool_
+        if (is_integer and arr.dtype.bitwidth < types.intp.bitwidth)\
+                or is_bool:
+            dtype = np.intp
+        else:
+            dtype = arr.dtype
 
+        def array_cumsum_impl(arr):
+            out = np.empty(arr.size, dtype)
+            c = np.zeros(1, dtype)[0]
```

Comment:
I believe this is a heap allocation which could be avoided? Would it be possible to use `c = scalar_dtype(0)` instead?